Debug/TimeProfiling.hpp

```diff
@@ -0,0 +1,208 @@
+ */
+
+
+#ifndef __TimeProfiling__
+#define __TimeProfiling__
+
+#include "Lib/Stack.hpp"
+#include "Lib/Option.hpp"
+#include "Kernel/Ordering.hpp"
+#include "Debug/Assertion.hpp"
+#include "Kernel/Term.hpp"
+#include <chrono>
+#include "Lib/MacroUtils.hpp"
+#define ENABLE_TIME_PROFILING 1
+
+namespace Shell {
+
+#if ENABLE_TIME_PROFILING
+
+/**
+ * Enables the runtime of the current block to be measured.
+ * Time traces are stored in a tree starting with a root node that measures the runtime of the whole
+ * probram execution, and subtrees that correspond to calls to TIME_TRACE.
+ * The parameter \param name is a human-readable name to be presented when the time trace is being
+ * outputed.
+ * If you want two blocks to contribute to the same node in the tree, they have to occur in the same
+ * subtree, and have to have the same const char* name. Note that just specifying the same literal
+ * twice does not guarantee that const char* are the same, even though that is the case for some
+ * compilers. Therefore one should use one static const char* for these kind of measurements.
+ *
+ * In addition to the time profiling tree a flattened version of the tree is being outputted at the
+ * end of a program execution, that presents the sum of all the runtimes a node was encountered on
+ * any branch of the tree.
+ *
+ * ```
+ */
+#define TIME_TRACE(name)                                                                            \
+  Shell::TimeTrace::ScopedTimer CONCAT_IDENTS(__time_trace_, __LINE__) (name);
+
+/**
+ * same as TIME_TRACE, but measures the time it takes for an expression to be evaluated. the
+ * arguments ... are the experssion to be time profiled.
+ */
+#define TIME_TRACE_EXPR(name, ...)                                                                  \
+  [&](){ TIME_TRACE(name); return __VA_ARGS__; }()
+
+/**
+ * sets a new node as the root of the time trace. this is useful when launching child process.
+ * The subtree will be set to the original root, when the call of TIME_TRACE_NEW_ROOT goes out of
+ * scope.  Therefore the statistics must be outputted before that to see any effect of this call.
+ */
+#define TIME_TRACE_NEW_ROOT(name)                                                                   \
+  TIME_TRACE(name)                                                                                  \
+  Shell::TimeTrace::ScopedChangeRoot CONCAT_IDENTS(__change_root_, __LINE__);
+
+#else // !ENABLE_TIME_PROFILING
+
+#define TIME_TRACE(name) {}
+#define TIME_TRACE_EXPR(name, ...) __VA_ARGS__
+#define TIME_TRACE_NEW_ROOT(name)
+
+#endif // ENABLE_TIME_PROFILING
+
+
+#if ENABLE_TIME_PROFILING
+
+
+/**
+ * A class to trace time for particular blocks. this class shall never be used directly,
+ * as it is preprocessed away when the flag ENABLE_TIME_PROFILING is set to false.
+ * Use the macros TIME_TRACE, TIME_TRACE_EXPR, and TIME_TRACE_NEW_ROOT instead.
+ */
+class TimeTrace
+{
+  using Clock = std::chrono::high_resolution_clock;
```

Comment:
[cppreference](https://en.cppreference.com/w/cpp/chrono/high_resolution_clock) maintains that `std::chrono::steady_clock` might be preferred here. Any thoughts?

Reply:
Agreed this should be changed!